client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java

```diff
@@ -0,0 +1,208 @@
+        boolean success = false;
+        boolean resendRequest = false;
+        ByteBuffer result = null;
+        Status status = Status.AVAILABLE_NOW;
+        //If bytesInBuffer is more than header buffer size, then issue a read request and get the event information. Else
+        // return empty buffer having status available later.
+        try {
+            headerReadingBuffer.clear();
+            int bytesInBuffer = input.bytesInBuffer();
+            //If outstanding request complete exceptionally then bytesInBuffer will return -1 and in read request will get the exception.
+            boolean isAvailable = bytesInBuffer == -1 || bytesInBuffer > headerReadingBuffer.capacity();
+            if (isAvailable  && (result = readEvent(firstByteTimeoutMillis)) != null) {
+                success = true;
+            } else {
+                status = Status.AVAILABLE_LATER;
+            }
+        } catch (TimeoutException e) {
+            resendRequest = true;
+            status = Status.AVAILABLE_LATER;
+            log.warn("Timeout observed while trying to read data from Segment store, the read request will be retransmitted");
+        } catch (EndOfSegmentException e) {
+            log.debug("Reached to end of segment for {}, end offset is {}.", segment, input.getOffset());
+            status = getStatus(e);
+        } catch (SegmentTruncatedException e) {
+            log.debug("Reached to end of segment for {}, end offset is {}.", segment, input.getOffset());
+            handleSegmentTruncated();
+            throw new TruncatedDataException();
+        } finally {
+            LoggerHelpers.traceLeave(log, "read", traceId, input.getSegmentId(), originalOffset, firstByteTimeoutMillis, success);
+            if (!success) {
+                // Reading failed, reset the offset to the original offset.
+                // The read request is retransmitted only in the case of a timeout.
+                input.setOffset(originalOffset, resendRequest);
+            }
+            input.fillBuffer();
+        }
+        return new EventReadWithStatusImpl<>(result == null ? null : deserializer.deserialize(result), status);
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        return input.fillBuffer().exceptionally(ex -> {
+            if (Exceptions.unwrap(ex) instanceof SegmentTruncatedException) {
+                result.completeExceptionally(new TruncatedDataException());
+            } else if (Exceptions.unwrap(ex) instanceof EndOfSegmentException) {
+                result.completeExceptionally(new EndOfSegmentException());
+            } else {
+                result.completeExceptionally(ex);
+            }
+            return null;
+        }).thenCompose(x -> {
+            //TODO : This check can be removed as fillBuffer itself ensure that once it will complete, there will be some data to read
```

Comment:
When the future completes there either will be data to read, or there will not because of truncation or end of segment. (Implying an answer has be obtained)